packages/driver/src/cypress/cy.ts

```diff
@@ -831,68 +829,76 @@ export class $Cy extends EventEmitter2 implements ITimeouts, IStability, IAssert
-    // - coming upon an assertion should only be
-    //   inserted if the previous command should
-    //   be replayed
+    const callback = (chainer, userInvocationStack, args) => {
+      // dont enqueue / inject any new commands if
+      // onInjectCommand returns false
+      const onInjectCommand = cy.state('onInjectCommand')
 
-    const commands = cy.getCommandsUntilFirstParentOrValidSubject(current)
+      if (_.isFunction(onInjectCommand) && onInjectCommand.call(cy, name, ...args) === false) {
+        return
+      }
 
-    if (commands) {
-      let initialCommand = commands.shift()
+      // Queries are functions that accept args (which is called once each time the command is used in the spec
+      // file), which return a function that accepts the subject (which is potentially called any number of times).
+      // The outer function is used to store any needed state needed by a particular instance of the command, such as
+      // a Cypress.log() instance, while the inner one (queryFn here) is the one that determines the next subject.
 
-      const commandsToInsert = _.reduce(commands, (memo, command, index) => {
-        const push = () => {
-          return memo.push(command)
-        }
+      // We enqueue the outer function as the "cypress command". See command_queue.ts for details on how this is
+      // invoked and the inner function retried.
+      const command = $Command.create({
+        name,
+        args,
+        type: 'dual',
+        chainerId: chainer.chainerId,
+        userInvocationStack,
+        query: true,
+        prevSubject: 'optional',
+      })
 
-        if (!(command.get('type') !== 'assertion')) {
-          // if we're an assertion and the prev command
-          // is in the memo, then push this one
-          if (memo.includes(command.get('prev'))) {
-            push()
-          }
-        } else if (!(command.get('subject') === initialCommand.get('subject'))) {
-          // when our subjects dont match then
-          // reset the initialCommand to this command
-          // so the next commands can compare against
-          // this one to figure out the changing subjects
-          initialCommand = command
-
-          push()
-        }
+      const cyFn = function (chainerId, ...args) {
+        return fn.apply(command, args)
+      }
+
+      cyFn.originalFn = fn
+      command.set('fn', cyFn)
 
-        return memo
-      }, [initialCommand])
+      cy.enqueue(command)
+    }
+
+    $Chainer.add(name, callback)
+
+    cy[name] = function (...args) {
+      cy.ensureRunnable(name)
+
+      // this is the first call on cypress
+      // so create a new chainer instance
+      const chainer = new $Chainer(cy.specWindow)
+
+      if (cy.state('chainerId')) {
+        cy.linkSubject(chainer.chainerId, cy.state('chainerId'))
+      }
 
-      const chainerId = this.state('chainerId')
+      const userInvocationStack = $stackUtils.captureUserInvocationStack(cy.specWindow.Error)
```

Comment:
I am not fully understanding what's going on here - why does `_addQuery` need to grab the userInvocationStack? Isn't this for error handling? Do we expect that here when they initially call `useQuery` (or is this to handle the cases where they used `useQuery` incorrectly?

Reply:
Ok, that explains how this works (had a vague idea, but this clarified). Last question would be why was this added here specifically in this PR? Did we move the logic here from somewhere else? Or did the `useQuery` addition require this to be added here?

Reply:
It's added here because `_addQuery` is a mirror to the existing `addCommand` - you can see the exact same line,

`const userInvocationStack = $stackUtils.captureUserInvocationStack(cy.specWindow.Error)`

up earlier in this file, line 791 inside addCommand.

`_addQuery` and `addCommand` are fairly similar in a lot of ways, but the details differ enough that I didn't want to mesh in a bunch of `if (isQuery) {...}` code branches into addCommand, and decided that some code duplication was the lesser of two evils.